### ISSUE TITLE
Does not work in atexit context

### ISSUE THREAD

**Issue by dtolnay** [dtolnay/syn#862]
```rust
extern "C" fn atexit_callback() {
    let _ = syn::parse_file("...");
}

fn main() {
    unsafe {
        libc::atexit(atexit_callback);
    }
}
```

This panicks while constructing a syn::Error parse error.

```console
thread '<unnamed>' panicked at 'use of std::thread::current() is not possible after the thread's local data has been destroyed', src/libstd/thread/mod.rs:643:5
stack backtrace:
   0: backtrace::backtrace::libunwind::trace
             at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.46/src/backtrace/libunwind.rs:86
   1: backtrace::backtrace::trace_unsynchronized
             at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.46/src/backtrace/mod.rs:66
   2: std::sys_common::backtrace::_print_fmt
             at src/libstd/sys_common/backtrace.rs:78
   3: <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt
             at src/libstd/sys_common/backtrace.rs:59
   4: core::fmt::write
             at src/libcore/fmt/mod.rs:1069
   5: std::io::Write::write_fmt
             at src/libstd/io/mod.rs:1504
   6: std::sys_common::backtrace::_print
             at src/libstd/sys_common/backtrace.rs:62
   7: std::sys_common::backtrace::print
             at src/libstd/sys_common/backtrace.rs:49
   8: std::panicking::default_hook::{{closure}}
             at src/libstd/panicking.rs:198
   9: std::panicking::default_hook
             at src/libstd/panicking.rs:218
  10: std::panicking::rust_panic_with_hook
             at src/libstd/panicking.rs:515
  11: rust_begin_unwind
             at src/libstd/panicking.rs:419
  12: core::panicking::panic_fmt
             at src/libcore/panicking.rs:111
  13: core::option::expect_failed
             at src/libcore/option.rs:1260
  14: core::option::Option<T>::expect
             at /rustc/c7087fe00d2ba919df1d813c040a5d47e43b0fe7/src/libcore/option.rs:347
  15: std::thread::current
             at src/libstd/thread/mod.rs:643
  16: syn::thread::ThreadBound<T>::new
             at ./.cargo/registry/src/github.com-1ecc6299db9ec823/syn-1.0.24/src/thread.rs:21
  17: syn::error::Error::new
             at ./.cargo/registry/src/github.com-1ecc6299db9ec823/syn-1.0.24/src/error.rs:139
  18: syn::error::new_at
             at ./.cargo/registry/src/github.com-1ecc6299db9ec823/syn-1.0.24/src/error.rs:248
  19: syn::lookahead::Lookahead1::error
             at ./.cargo/registry/src/github.com-1ecc6299db9ec823/syn-1.0.24/src/lookahead.rs:132
  20: syn::item::parsing::<impl syn::parse::Parse for syn::item::Item>::parse
             at ./.cargo/registry/src/github.com-1ecc6299db9ec823/syn-1.0.24/src/item.rs:1315
  21: syn::parse::ParseBuffer::parse
             at ./.cargo/registry/src/github.com-1ecc6299db9ec823/syn-1.0.24/src/parse.rs:449
  22: syn::file::parsing::<impl syn::parse::Parse for syn::file::File>::parse
             at ./.cargo/registry/src/github.com-1ecc6299db9ec823/syn-1.0.24/src/file.rs:91
  23: core::ops::function::FnOnce::call_once
             at /rustc/c7087fe00d2ba919df1d813c040a5d47e43b0fe7/src/libcore/ops/function.rs:232
  24: <F as syn::parse::Parser>::parse2
             at ./.cargo/registry/src/github.com-1ecc6299db9ec823/syn-1.0.24/src/parse.rs:1167
  25: syn::parse::Parser::parse_str
             at ./.cargo/registry/src/github.com-1ecc6299db9ec823/syn-1.0.24/src/parse.rs:1132
  26: syn::parse_str
             at ./.cargo/registry/src/github.com-1ecc6299db9ec823/syn-1.0.24/src/lib.rs:894
  27: syn::parse_file
             at ./.cargo/registry/src/github.com-1ecc6299db9ec823/syn-1.0.24/src/lib.rs:953
  28: playground::atexit_callback
             at src/main.rs:2
  29: <unknown>
  30: exit
  31: __libc_start_main
  32: _start
```

**Comment by taiki-e** [comment]
Probably related: https://github.com/rust-lang/rust/issues/28129

**Comment by dtolnay** [comment]
I implemented a fix in #1317 but it's fairly troublesome, so I think this can just remain unsupported.
